### ISSUE TITLE
Classloading issue when using appenders outside the log4j-core-2.17.1.jar

### ISSUE THREAD

**Issue by nieder-github** [ot4i/node-for-log4j#3]
There is a known problem when using additional classes / appender / custom extensions to log4j outside the ones in the par file.
A case has been opened to IBM.

A workaround is:
1) unzip the par file 
2) copy all jars from the lib folder to the shared-classes folder 
3) jar everything in the classes folder as Log4jLoggingNode_v2.0.0.jar
4) copy the new created Log4jLoggingNode_v2.0.0.jar to the shared-classes folder

**Comment by mojahidam** [comment]
I have tried the workaround mentioned here. Now I have below 4 bars in my `/var/mqsi/shared-classes` directory.

```
-rw-rw-r--. 1 aceuser aceuser   65261 Jul 20 21:59 jakarta-oro-2.0.8.jar
-rw-rw-r--. 1 aceuser aceuser 1790452 Jul 20 21:59 log4j-core-2.17.1.jar
-rw-rw-r--. 1 aceuser aceuser  301873 Jul 20 21:59 log4j-api-2.17.1.jar
-rw-rw-r--. 1 aceuser aceuser   16383 Jul 20 22:43 Log4jLoggingNode_v2.0.0.jar
```

I have deployed my bar file successfully. When I try to run some transaction, it is creating the `.log` file but is unable to log anything.

Below are my procedures and functions declarations.
```
BROKER SCHEMA Logger

DECLARE rc SHARED BOOLEAN;
DECLARE UDP_LogConfig EXTERNAL CHARACTER '';

CREATE PROCEDURE initializeLog4j(IN file CHARACTER)
BEGIN
	IF ( rc IS NULL OR rc = FALSE ) THEN
		CALL initLog4j(file) INTO rc;
		CALL logDebug('Logger Initialization Successful.');
	END IF;
END;

CREATE PROCEDURE logDebug(IN text CHARACTER)
BEGIN
	CALL initializeLog4j(UDP_LogConfig);
	CALL log4j(MessageFlowLabel, NodeLabel, 'DEBUG', text) INTO rc;
END;

CREATE PROCEDURE logInfo(IN text CHARACTER)
BEGIN
	CALL initializeLog4j(UDP_LogConfig);
	CALL log4j(MessageFlowLabel, NodeLabel, 'INFO', text) INTO rc;
END;

CREATE PROCEDURE logWarn(IN text CHARACTER)
BEGIN
	CALL initializeLog4j(UDP_LogConfig);
	CALL log4j(MessageFlowLabel, NodeLabel, 'WARN', text) INTO rc;
END;

CREATE PROCEDURE logError(IN text CHARACTER)
BEGIN
	CALL initializeLog4j(UDP_LogConfig);
	CALL log4j(MessageFlowLabel, NodeLabel, 'ERROR', text) INTO rc;
END;

/**
* Declaration from Java Method Invocation
*
* Do Not Modify
*
*/

CREATE FUNCTION initLog4j( IN CONFIG_FILE_NAME CHARACTER )
RETURNS BOOLEAN
LANGUAGE JAVA
EXTERNAL NAME "com.ibm.broker.IAM3.Log4jNode.initLog4j";

CREATE FUNCTION log4j( IN COMPONENT_NAME CHARACTER, IN LOGGER_NAME CHARACTER, IN LEVEL CHARACTER, IN TEXT CHARACTER )
RETURNS BOOLEAN
LANGUAGE JAVA
EXTERNAL NAME "com.ibm.broker.IAM3.Log4jNode.log";
```

Below is how I call the log procedures.
```
CALL Logger.logDebug('This is a debug message');
CALL Logger.logInfo('This is an info message');
CALL Logger.logInfo('This is a warn message');
CALL Logger.logInfo('This is an error message');
```

Here is my `log4j.xml` configuration file.
```
<?xml version="1.0" encoding="UTF-8"?>
<Configuration monitorInterval="10" strict="false" packages="com.ibm.log4j.filter">
	<Appenders>
		<File name="FILE" fileName="/home/aceuser/ACE/DEV/INDITEX/INBOUND/Log4j2.log" immediateFlush="true" append="true">
			<PatternLayout pattern="%d{yyy-MM-dd HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n"/>
			<ThresholdFilter level="DEBUG"/>
		</File>
	</Appenders>
	<Loggers>
		<Root>
			<AppenderRef ref="FILE"/>
		</Root>
	</Loggers>
</Configuration>
```

See it is able to create the `.log` file but is unable to log anything.
```
[aceuser@tuat9tgb01 INBOUND]$ ls -lrt
total 8
-rw-rw----. 1 aceuser aceuser    0 Jul 20 22:47 Log4j2.log
```

**Comment by nieder-github** [comment]
I'd like you to enable log4j tracing as described in the log4j docu:
Setting status="trace" is one of the first tools available to you if you need to troubleshoot log4j.
-> 
<Configuration status="trace" .....
</Configuration>
(Alternatively, setting system property log4j2.debug will also print internal Log4j2 logging to the console, including internal logging that took place before the configuration file was found.)

**Comment by mojahidam** [comment]
I have enabled trace and captured the log from stdout. See below.

Config File
```
<?xml version="1.0" encoding="UTF-8"?>
<Configuration status="trace" monitorInterval="10" strict="false" packages="com.ibm.log4j.filter">
	<Appenders>
		<File name="FILE" fileName="/home/aceuser/ACE/DEV/INDITEX/log/Log4j2.log" immediateFlush="true" append="true">
			<PatternLayout pattern="%d{yyy-MM-dd HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n"/>
			<ThresholdFilter level="DEBUG"/>
		</File>
	</Appenders>
	<Loggers>
		<Root>
			<AppenderRef ref="FILE"/>
		</Root>
	</Loggers>
</Configuration>
```
Stdout Log
```
2022-08-11 10:29:01.591     44 2022-08-11 10:29:01,591 Get Stub Data DEBUG Apache Log4j Core 2.17.1 initializing configuration XmlConfiguration[location=/home/aceuser/ACE/DEV/INDITEX/config/log4j.xml]
2022-08-11 10:29:01.600     44 2022-08-11 10:29:01,600 Get Stub Data DEBUG Installed 1 script engine
2022-08-11 10:29:01.835     44 2022-08-11 10:29:01,834 Get Stub Data DEBUG Oracle Nashorn version: 1.8.0, language: ECMAScript, threading: Not Thread Safe, compile: true, names: [nashorn, Nashorn, js, JS, JavaScript, javascript, ECMAScript, ecmascript], factory class: jdk.nashorn.api.scripting.NashornScriptEngineFactory
2022-08-11 10:29:01.842     44 2022-08-11 10:29:01,842 Get Stub Data DEBUG Took 0.005065 seconds to load 1 plugins from package com.ibm.log4j.filter
2022-08-11 10:29:01.844     44 2022-08-11 10:29:01,844 Get Stub Data DEBUG PluginManager 'Core' found 128 plugins
2022-08-11 10:29:01.845     44 2022-08-11 10:29:01,845 Get Stub Data DEBUG PluginManager 'Level' found 0 plugins
2022-08-11 10:29:01.853     44 2022-08-11 10:29:01,852 Get Stub Data DEBUG PluginManager 'Lookup' found 16 plugins
2022-08-11 10:29:01.856     44 2022-08-11 10:29:01,856 Get Stub Data DEBUG Building Plugin[name=layout, class=org.apache.logging.log4j.core.layout.PatternLayout].
2022-08-11 10:29:01.881     44 2022-08-11 10:29:01,881 Get Stub Data TRACE TypeConverterRegistry initializing.
2022-08-11 10:29:01.892     44 2022-08-11 10:29:01,892 Get Stub Data DEBUG PluginManager 'TypeConverter' found 26 plugins
2022-08-11 10:29:01.917     44 2022-08-11 10:29:01,917 Get Stub Data DEBUG PatternLayout$Builder(pattern="%d{yyy-MM-dd HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n", PatternSelector=null, Configuration(/home/aceuser/ACE/DEV/INDITEX/config/log4j.xml), Replace=null, charset="null", alwaysWriteExceptions="null", disableAnsi="null", noConsoleNoAnsi="null", header="null", footer="null")
2022-08-11 10:29:01.919     44 2022-08-11 10:29:01,919 Get Stub Data DEBUG PluginManager 'Converter' found 45 plugins
2022-08-11 10:29:01.943     44 2022-08-11 10:29:01,942 Get Stub Data DEBUG Building Plugin[name=filter, class=org.apache.logging.log4j.core.filter.ThresholdFilter].
2022-08-11 10:29:01.968     44 2022-08-11 10:29:01,967 Get Stub Data DEBUG createFilter(level="DEBUG", onMatch="null", onMismatch="null")
2022-08-11 10:29:01.969     44 2022-08-11 10:29:01,969 Get Stub Data DEBUG Building Plugin[name=appender, class=org.apache.logging.log4j.core.appender.FileAppender].
2022-08-11 10:29:02.004     44 2022-08-11 10:29:02,004 Get Stub Data DEBUG FileAppender$Builder(fileName="/home/aceuser/ACE/DEV/INDITEX/log/Log4j2.log", append="true", locking="null", advertise="null", advertiseUri="null", createOnDemand="null", filePermissions="null", fileOwner="null", fileGroup="null", bufferedIo="null", bufferSize="null", immediateFlush="true", ignoreExceptions="null", PatternLayout(%d{yyy-MM-dd HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n), name="FILE", Configuration(/home/aceuser/ACE/DEV/INDITEX/config/log4j.xml), ThresholdFilter(DEBUG), ={})
2022-08-11 10:29:02.008     44 2022-08-11 10:29:02,007 Get Stub Data DEBUG Starting FileManager /home/aceuser/ACE/DEV/INDITEX/log/Log4j2.log
2022-08-11 10:29:02.009     44 2022-08-11 10:29:02,009 Get Stub Data DEBUG Building Plugin[name=appenders, class=org.apache.logging.log4j.core.config.AppendersPlugin].
2022-08-11 10:29:02.011     44 2022-08-11 10:29:02,010 Get Stub Data DEBUG createAppenders(={FILE})
2022-08-11 10:29:02.012     44 2022-08-11 10:29:02,012 Get Stub Data DEBUG Building Plugin[name=AppenderRef, class=org.apache.logging.log4j.core.config.AppenderRef].
2022-08-11 10:29:02.014     44 2022-08-11 10:29:02,014 Get Stub Data DEBUG createAppenderRef(ref="FILE", level="null", Filter=null)
2022-08-11 10:29:02.015     44 2022-08-11 10:29:02,015 Get Stub Data DEBUG Building Plugin[name=root, class=org.apache.logging.log4j.core.config.LoggerConfig$RootLogger].
2022-08-11 10:29:02.017     44 2022-08-11 10:29:02,017 Get Stub Data DEBUG createLogger(additivity="null", level="null", includeLocation="null", ={FILE}, ={}, Configuration(/home/aceuser/ACE/DEV/INDITEX/config/log4j.xml), Filter=null)
2022-08-11 10:29:02.021     44 2022-08-11 10:29:02,021 Get Stub Data DEBUG Building Plugin[name=loggers, class=org.apache.logging.log4j.core.config.LoggersPlugin].
2022-08-11 10:29:02.022     44 2022-08-11 10:29:02,022 Get Stub Data DEBUG createLoggers(={root})
2022-08-11 10:29:02.024     44 2022-08-11 10:29:02,024 Get Stub Data DEBUG Configuration XmlConfiguration[location=/home/aceuser/ACE/DEV/INDITEX/config/log4j.xml] initialized
2022-08-11 10:29:02.025     44 2022-08-11 10:29:02,024 Get Stub Data DEBUG Starting configuration XmlConfiguration[location=/home/aceuser/ACE/DEV/INDITEX/config/log4j.xml]
2022-08-11 10:29:02.026     44 2022-08-11 10:29:02,026 Get Stub Data DEBUG Log4j2 ConfigurationScheduler starting 1 threads
2022-08-11 10:29:02.029     44 2022-08-11 10:29:02,028 Get Stub Data DEBUG Started configuration XmlConfiguration[location=/home/aceuser/ACE/DEV/INDITEX/config/log4j.xml] OK.
2022-08-11 10:29:02.030     44 2022-08-11 10:29:02,030 Get Stub Data TRACE Stopping org.apache.logging.log4j.core.config.DefaultConfiguration@4c781f26...
2022-08-11 10:29:02.031     44 2022-08-11 10:29:02,031 Get Stub Data TRACE DefaultConfiguration notified 1 ReliabilityStrategies that config will be stopped.
2022-08-11 10:29:02.032     44 2022-08-11 10:29:02,032 Get Stub Data TRACE DefaultConfiguration stopping root LoggerConfig.
2022-08-11 10:29:02.033     44 2022-08-11 10:29:02,033 Get Stub Data TRACE DefaultConfiguration notifying ReliabilityStrategies that appenders will be stopped.
2022-08-11 10:29:02.034     44 2022-08-11 10:29:02,034 Get Stub Data TRACE DefaultConfiguration stopping remaining Appenders.
2022-08-11 10:29:02.036     44 2022-08-11 10:29:02,035 Get Stub Data DEBUG Shutting down OutputStreamManager SYSTEM_OUT.false.false-1
2022-08-11 10:29:02.036     44 2022-08-11 10:29:02,036 Get Stub Data DEBUG OutputStream closed
2022-08-11 10:29:02.037     44 2022-08-11 10:29:02,037 Get Stub Data DEBUG Shut down OutputStreamManager SYSTEM_OUT.false.false-1, all resources released: true
2022-08-11 10:29:02.038     44 2022-08-11 10:29:02,038 Get Stub Data DEBUG Appender DefaultConsole-1 stopped with status true
2022-08-11 10:29:02.040     44 2022-08-11 10:29:02,039 Get Stub Data TRACE DefaultConfiguration stopped 1 remaining Appenders.
2022-08-11 10:29:02.041     44 2022-08-11 10:29:02,040 Get Stub Data TRACE DefaultConfiguration cleaning Appenders from 1 LoggerConfigs.
2022-08-11 10:29:02.042     44 2022-08-11 10:29:02,041 Get Stub Data DEBUG Stopped org.apache.logging.log4j.core.config.DefaultConfiguration@4c781f26 OK
2022-08-11 10:29:02.124     44 2022-08-11 10:29:02,124 Get Stub Data TRACE Reregistering MBeans after reconfigure. Selector=org.apache.logging.log4j.core.selector.ClassLoaderContextSelector@1b45a61d
2022-08-11 10:29:02.125     44 2022-08-11 10:29:02,125 Get Stub Data TRACE Reregistering context (1/1): 'd2882113' org.apache.logging.log4j.core.LoggerContext@7c34b521
2022-08-11 10:29:02.127     44 2022-08-11 10:29:02,126 Get Stub Data TRACE Unregistering but no MBeans found matching 'org.apache.logging.log4j2:type=d2882113'
2022-08-11 10:29:02.128     44 2022-08-11 10:29:02,128 Get Stub Data TRACE Unregistering but no MBeans found matching 'org.apache.logging.log4j2:type=d2882113,component=StatusLogger'
2022-08-11 10:29:02.129     44 2022-08-11 10:29:02,128 Get Stub Data TRACE Unregistering but no MBeans found matching 'org.apache.logging.log4j2:type=d2882113,component=ContextSelector'
2022-08-11 10:29:02.130     44 2022-08-11 10:29:02,130 Get Stub Data TRACE Unregistering but no MBeans found matching 'org.apache.logging.log4j2:type=d2882113,component=Loggers,name=*'
2022-08-11 10:29:02.131     44 2022-08-11 10:29:02,131 Get Stub Data TRACE Unregistering but no MBeans found matching 'org.apache.logging.log4j2:type=d2882113,component=Appenders,name=*'
2022-08-11 10:29:02.132     44 2022-08-11 10:29:02,131 Get Stub Data TRACE Unregistering but no MBeans found matching 'org.apache.logging.log4j2:type=d2882113,component=AsyncAppenders,name=*'
2022-08-11 10:29:02.133     44 2022-08-11 10:29:02,133 Get Stub Data TRACE Unregistering but no MBeans found matching 'org.apache.logging.log4j2:type=d2882113,component=AsyncLoggerRingBuffer'
2022-08-11 10:29:02.134     44 2022-08-11 10:29:02,134 Get Stub Data TRACE Unregistering but no MBeans found matching 'org.apache.logging.log4j2:type=d2882113,component=Loggers,name=*,subtype=RingBuffer'
2022-08-11 10:29:02.137     44 2022-08-11 10:29:02,137 Get Stub Data DEBUG Registering MBean org.apache.logging.log4j2:type=d2882113
2022-08-11 10:29:02.140     44 2022-08-11 10:29:02,140 Get Stub Data DEBUG Registering MBean org.apache.logging.log4j2:type=d2882113,component=StatusLogger
2022-08-11 10:29:02.142     44 2022-08-11 10:29:02,142 Get Stub Data DEBUG Registering MBean org.apache.logging.log4j2:type=d2882113,component=ContextSelector
2022-08-11 10:29:02.145     44 2022-08-11 10:29:02,144 Get Stub Data DEBUG Registering MBean org.apache.logging.log4j2:type=d2882113,component=Loggers,name=
2022-08-11 10:29:02.147     44 2022-08-11 10:29:02,147 Get Stub Data DEBUG Registering MBean org.apache.logging.log4j2:type=d2882113,component=Appenders,name=FILE
2022-08-11 10:29:02.149     44 2022-08-11 10:29:02,149 Get Stub Data TRACE Using default SystemClock for timestamps.
2022-08-11 10:29:02.150     44 2022-08-11 10:29:02,150 Get Stub Data DEBUG org.apache.logging.log4j.core.util.SystemClock does not support precise timestamps.
2022-08-11 10:29:02.151     44 2022-08-11 10:29:02,151 Get Stub Data TRACE Using DummyNanoClock for nanosecond timestamps.
2022-08-11 10:29:02.152     44 2022-08-11 10:29:02,152 Get Stub Data DEBUG Reconfiguration complete for context[name=d2882113] at URI /home/aceuser/ACE/DEV/INDITEX/config/log4j.xml (org.apache.logging.log4j.core.LoggerContext@7c34b521) with optional ClassLoader: null
2022-08-11 10:29:02.153     44 2022-08-11 10:29:02,153 Get Stub Data DEBUG Shutdown hook enabled. Registering a new one.
2022-08-11 10:29:02.155     44 2022-08-11 10:29:02,155 Get Stub Data DEBUG LoggerContext[name=d2882113, org.apache.logging.log4j.core.LoggerContext@7c34b521] started OK.
2022-08-11 10:29:12.029     56 2022-08-11 10:29:12,028 Log4j2-TF-2-Scheduled-1 TRACE WatchRunnable run triggered.
2022-08-11 10:29:12.030     56 2022-08-11 10:29:12,030 Log4j2-TF-2-Scheduled-1 TRACE WatchRunnable run ended.
```

**Comment by nieder-github** [comment]
Please add the level attribute to the Root logger and see if that helps.
<Root level="debug">
....
</Root>

**Comment by mojahidam** [comment]
It worked. Thanks!
```
<?xml version="1.0" encoding="UTF-8"?>
<Configuration status="trace" monitorInterval="10" strict="false" packages="com.ibm.log4j.filter">
	<Appenders>
		<File name="FILE" fileName="/home/aceuser/ACE/DEV/INDITEX/log/Log4j2.log" immediateFlush="true" append="true">
			<PatternLayout pattern="%d{yyy-MM-dd HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n"/>
			<ThresholdFilter level="DEBUG"/>
		</File>
	</Appenders>
	<Loggers>
		<Root level="debug">
			<AppenderRef ref="FILE"/>
		</Root>
	</Loggers>
</Configuration>
```

**Comment by nagkalgur** [comment]
which IIB version are you using.

**Comment by nieder-github** [comment]
See requirements: at least ACE v11.0.0.3 or v12.0.3.0
On z/OS IIB v10.0.0.26 should also work.

**Comment by nieder-github** [comment]
solved with version 2.1.0:
define a pluginFolder as policy ( ACE ) or Configurable Service ( IIB ) for your extensions
